### ISSUE TITLE
recently gpodder started to download the same soundcloud episodes over and over again

### ISSUE THREAD

**Issue by christofdamian** [gpodder/gpodder#1064]
This seems to just affect soundcloud podcasts. 
I also tried setting them manually as "old", but this didn't help either. 
I am pretty sure this only started in the recent few weeks. Maybe Since May. 

I am on gPodder 3.10.19 (2021-04-15) using the Fedora package.

**Comment by auouymous** [comment]
Can you provide a feed URL for testing?

**Comment by christofdamian** [comment]
> Can you provide a feed URL for testing?

Yes, sorry. For example:
* http://soundcloud.com/motik
* https://soundcloud.com/chris-liebing
* http://soundcloud.com/richiehawtin

but as far as I can see all of my soundcloud subscriptions are affected. This is my full list of all podcasts: https://gpodder.net/user/cdamian/subscriptions (yes, they are all electronic music)

**Comment by auouymous** [comment]
I subscribed to motik, marked the episodes as old and updated the feed which did not try to download them again. Does it happen for you with a short interval between updates? Could you try running gpodder in verbose mode (`gpodder -v` on the command line) and then look in the logs for any issues after an update of one feed?

**Comment by christofdamian** [comment]
I usually only run and check every few days. This is the log from just manually updating motik, downloading and the updating again:

```
1623063515.555529 [gpodder.util] DEBUG: run_in_background: <function gPodder.update_feed_cache.<locals>.update_feed_cache_proc at 0x7ff317d7caf0> (False)
1623063515.558671 [gpodder.registry] INFO: feed_handler resolved by fetch_channel from gpodder.plugins.soundcloud: None -> <gpodder.feedcore.Result object at 0x7ff317b7b520>
1623063515.558924 [gpodder.plugins.soundcloud] DEBUG: loading https://api.soundcloud.com/users/901720/tracks.json?consumer_key=zrweghtEtnZLpXf3mlm8mQ&limit=200
1623063515.559995 [urllib3.connectionpool] DEBUG: Starting new HTTPS connection (1): api.soundcloud.com:443
1623063515.593628 [gpodder.gtkui.main] INFO: Updating motik on Soundcloud (1/1)
1623063515.607155 [urllib3.connectionpool] DEBUG: https://api.soundcloud.com:443 "GET /users/901720/tracks.json?consumer_key=zrweghtEtnZLpXf3mlm8mQ&limit=200 HTTP/1.1" 200 1605
1623063515.608076 [gpodder.plugins.soundcloud] INFO: 2/2 downloadable tracks for user motik tracks feed
1623063515.621224 [gpodder.dbsqlite] DEBUG: Purge requested for podcast 19
1623063515.621446 [gpodder.dbsqlite] DEBUG: Commit.
1623063515.633099 [gpodder.gtkui.services] DEBUG: cover download request for http://soundcloud.com/motik
1623063515.633184 [gpodder.util] DEBUG: run_in_background: <function CoverDownloader.request_cover.<locals>.<lambda> at 0x7ff31447c9d0> (False)
1623063515.633556 [gpodder.coverart] INFO: Downloading cover art: https://i1.sndcdn.com/avatars-000001146600-xeho16-large.jpg
1623063515.635422 [urllib3.connectionpool] DEBUG: Starting new HTTPS connection (1): i1.sndcdn.com:443
1623063515.645010 [gpodder.gtkui.main] DEBUG: Updated channel is active, updating UI
1623063515.669825 [gpodder.my] DEBUG: Processing received episode actions...
1623063515.674567 [gpodder.my] DEBUG: Received episode actions processed.
1623063515.674667 [gpodder.dbsqlite] DEBUG: Commit.
1623063515.681282 [urllib3.connectionpool] DEBUG: https://i1.sndcdn.com:443 "GET /avatars-000001146600-xeho16-large.jpg HTTP/1.1" 404 21
1623063515.682073 [gpodder.coverart] WARNING: Cover art download failed: https://i1.sndcdn.com/avatars-000001146600-xeho16-large.jpg returned status code 404
1623063515.697134 [gpodder.gtkui.model] INFO: Clearing cover from cache: http://soundcloud.com/motik




1623063619.636453 [gpodder.gtkui.main] DEBUG: Downloading episode: Motik - Dunkelschunkel
1623063619.637096 [gpodder.model] INFO: Setting download filename: Motik - Dunkelschunkel.mp3
1623063619.637725 [gpodder.gtkui.main] DEBUG: Downloading episode: Motik 121212121212 mp3
1623063619.637916 [gpodder.model] INFO: Setting download filename: Motik 121212121212 mp3.mp3
1623063619.639010 [gpodder.my] DEBUG: Storing 1 episode download actions
1623063619.639459 [gpodder.download] INFO: 1 tasks to do, can start at most 5 threads
1623063619.639538 [gpodder.download] INFO: Starting new worker thread.
1623063619.639599 [gpodder.util] DEBUG: run_in_background: <bound method DownloadQueueWorker.run of MainThread> (False)
1623063619.639786 [gpodder.download] INFO: Starting new thread: Thread-43
1623063619.639934 [gpodder.my] DEBUG: Storing 1 episode download actions
1623063619.640239 [gpodder.download] INFO: Thread-43 is processing: Motik - Dunkelschunkel
1623063619.640471 [gpodder.download] INFO: Downloading https://api.soundcloud.com/tracks/2564505/stream?consumer_key=zrweghtEtnZLpXf3mlm8mQ
1623063619.640657 [gpodder.download] INFO: 1 tasks to do, can start at most 4 threads
1623063619.640932 [gpodder.download] INFO: Starting new worker thread.
1623063619.641734 [gpodder.util] DEBUG: run_in_background: <bound method DownloadQueueWorker.run of MainThread> (False)
1623063619.641918 [urllib3.connectionpool] DEBUG: Starting new HTTPS connection (1): api.soundcloud.com:443
1623063619.642014 [gpodder.download] INFO: Starting new thread: Thread-44
1623063619.642668 [gpodder.download] INFO: Thread-44 is processing: Motik 121212121212 mp3
1623063619.642955 [gpodder.download] INFO: Downloading https://api.soundcloud.com/tracks/71552595/stream?consumer_key=zrweghtEtnZLpXf3mlm8mQ
1623063619.644996 [urllib3.connectionpool] DEBUG: Starting new HTTPS connection (1): api.soundcloud.com:443
1623063619.650986 [gpodder.my] DEBUG: Flush requested, already waiting.
1623063619.998711 [urllib3.connectionpool] DEBUG: https://api.soundcloud.com:443 "GET /tracks/2564505/stream?consumer_key=zrweghtEtnZLpXf3mlm8mQ HTTP/1.1" 302 569
1623063620.000257 [urllib3.connectionpool] DEBUG: Starting new HTTPS connection (1): cf-media.sndcdn.com:443
1623063620.008659 [urllib3.connectionpool] DEBUG: https://api.soundcloud.com:443 "GET /tracks/71552595/stream?consumer_key=zrweghtEtnZLpXf3mlm8mQ HTTP/1.1" 302 571
1623063620.011187 [urllib3.connectionpool] DEBUG: Starting new HTTPS connection (1): cf-media.sndcdn.com:443
1623063620.146719 [urllib3.connectionpool] DEBUG: https://cf-media.sndcdn.com:443 "GET /f5l5Gl2bmavl.128.mp3?Policy=eyJTdGF0ZW1lbnQiOlt7IlJlc291cmNlIjoiKjovL2NmLW1lZGlhLnNuZGNkbi5jb20vZjVsNUdsMmJtYXZsLjEyOC5tcDMiLCJDb25kaXRpb24iOnsiRGF0ZUxlc3NUaGFuIjp7IkFXUzpFcG9jaFRpbWUiOjE2MjMwNjk4OTF9fX1dfQ__&Signature=aeth~os5wqLO7OvpmXI1ya6hFZ01jm-vHdbCOohDFJ~Zimsxze3WAlF3nUiQyXW49CxEV-WXPFV7S4ixXeMTPzWy9Jn2Nl6Zi8Odf9DIoxyNxPAo7f7GQ~0pnLGGTF0BQ5~CqWmT8lo3v8nyK~JGDFCjKRBRe-sWA92VpUszhnimyll3JFDoMXdAF~kJjImeSRXJr7xTLVS3P104zNRana~yhTk0k8dRfcStaUQXlztAzB9HUn7Jd85x4cqos3GLYGR~mPho0U2n3w6af6XFyaGYhNeORmTiZUFt5peXAFLx2SYXs86tPJy0EUCMQu9h8~-4SxZ1GITlpA2VxiygkQ__&Key-Pair-Id=APKAI6TU7MMXM5DG6EPQ HTTP/1.1" 200 98380903
1623063620.152521 [urllib3.connectionpool] DEBUG: https://cf-media.sndcdn.com:443 "GET /QX7dz9zhSE2m.128.mp3?Policy=eyJTdGF0ZW1lbnQiOlt7IlJlc291cmNlIjoiKjovL2NmLW1lZGlhLnNuZGNkbi5jb20vUVg3ZHo5emhTRTJtLjEyOC5tcDMiLCJDb25kaXRpb24iOnsiRGF0ZUxlc3NUaGFuIjp7IkFXUzpFcG9jaFRpbWUiOjE2MjMwNzE3MTB9fX1dfQ__&Signature=TADMlNl6gsn~RZplb-DG47ZHQahs2Gk37KZI6iFZ1UaefXQv1WOt9FEjdJE6db9hsg5Uyp~6AXX1ayj7XOmVXLCVLe-aVYmwFyJOuk9eCd8YKvJvqyz~C~aoUXc-bzEgnISim2V3eUJSxb0d7A1BpL8a1sh5PUrt95fa2UTnkPbPS8-tpmc46dRNzopZoI8lksh0ydZYUf8y~nWX3jqCSFwutbcnytOWfxPdEQ9GADj2oc-JmrK4vS10a6qnSduZWjYE5naCHeqddazxcfNRwpbuL5XVb~2eg6EfksiH-PNQ1lVvIjIc6okyqqtTAdLoD3lhOSfXynqx5D8TKtmUrQ__&Key-Pair-Id=APKAI6TU7MMXM5DG6EPQ HTTP/1.1" 200 127518092
1623063622.662215 [gpodder.model] INFO: Choosing new filename: /home/cdamian/Music/gpodder-downloads/motik on Soundcloud/f5l5Gl2bmavl.128.mp3
1623063622.662377 [gpodder.model] INFO: Updating filename of https://api.soundcloud.com/tracks/2564505/stream?consumer_key=zrweghtEtnZLpXf3mlm8mQ to "f5l5Gl2bmavl.128.mp3".
1623063622.662694 [gpodder.download] INFO: Download was redirected (https://cf-media.sndcdn.com/f5l5Gl2bmavl.128.mp3?Policy=eyJTdGF0ZW1lbnQiOlt7IlJlc291cmNlIjoiKjovL2NmLW1lZGlhLnNuZGNkbi5jb20vZjVsNUdsMmJtYXZsLjEyOC5tcDMiLCJDb25kaXRpb24iOnsiRGF0ZUxlc3NUaGFuIjp7IkFXUzpFcG9jaFRpbWUiOjE2MjMwNjk4OTF9fX1dfQ__&Signature=aeth~os5wqLO7OvpmXI1ya6hFZ01jm-vHdbCOohDFJ~Zimsxze3WAlF3nUiQyXW49CxEV-WXPFV7S4ixXeMTPzWy9Jn2Nl6Zi8Odf9DIoxyNxPAo7f7GQ~0pnLGGTF0BQ5~CqWmT8lo3v8nyK~JGDFCjKRBRe-sWA92VpUszhnimyll3JFDoMXdAF~kJjImeSRXJr7xTLVS3P104zNRana~yhTk0k8dRfcStaUQXlztAzB9HUn7Jd85x4cqos3GLYGR~mPho0U2n3w6af6XFyaGYhNeORmTiZUFt5peXAFLx2SYXs86tPJy0EUCMQu9h8~-4SxZ1GITlpA2VxiygkQ__&Key-Pair-Id=APKAI6TU7MMXM5DG6EPQ). New filename: f5l5Gl2bmavl.128.mp3
1623063622.663513 [rename_download] INFO: Renaming: /home/cdamian/Music/gpodder-downloads/motik on Soundcloud/f5l5Gl2bmavl.128.mp3 -> /home/cdamian/Music/gpodder-downloads/motik on Soundcloud/Motik - Dunkelschunkel.mp3
1623063622.663752 [gpodder.dbsqlite] DEBUG: Commit.
1623063622.720710 [tagging] INFO: tagging.on_episode_downloaded(motik on Soundcloud/Motik - Dunkelschunkel)
1623063622.721103 [gpodder.download] INFO: No more tasks for Thread-43 to carry out.
1623063623.170276 [gpodder.model] INFO: Choosing new filename: /home/cdamian/Music/gpodder-downloads/motik on Soundcloud/QX7dz9zhSE2m.128.mp3
1623063623.170353 [gpodder.model] INFO: Updating filename of https://api.soundcloud.com/tracks/71552595/stream?consumer_key=zrweghtEtnZLpXf3mlm8mQ to "QX7dz9zhSE2m.128.mp3".
1623063623.170579 [gpodder.download] INFO: Download was redirected (https://cf-media.sndcdn.com/QX7dz9zhSE2m.128.mp3?Policy=eyJTdGF0ZW1lbnQiOlt7IlJlc291cmNlIjoiKjovL2NmLW1lZGlhLnNuZGNkbi5jb20vUVg3ZHo5emhTRTJtLjEyOC5tcDMiLCJDb25kaXRpb24iOnsiRGF0ZUxlc3NUaGFuIjp7IkFXUzpFcG9jaFRpbWUiOjE2MjMwNzE3MTB9fX1dfQ__&Signature=TADMlNl6gsn~RZplb-DG47ZHQahs2Gk37KZI6iFZ1UaefXQv1WOt9FEjdJE6db9hsg5Uyp~6AXX1ayj7XOmVXLCVLe-aVYmwFyJOuk9eCd8YKvJvqyz~C~aoUXc-bzEgnISim2V3eUJSxb0d7A1BpL8a1sh5PUrt95fa2UTnkPbPS8-tpmc46dRNzopZoI8lksh0ydZYUf8y~nWX3jqCSFwutbcnytOWfxPdEQ9GADj2oc-JmrK4vS10a6qnSduZWjYE5naCHeqddazxcfNRwpbuL5XVb~2eg6EfksiH-PNQ1lVvIjIc6okyqqtTAdLoD3lhOSfXynqx5D8TKtmUrQ__&Key-Pair-Id=APKAI6TU7MMXM5DG6EPQ). New filename: QX7dz9zhSE2m.128.mp3
1623063623.171009 [rename_download] INFO: Renaming: /home/cdamian/Music/gpodder-downloads/motik on Soundcloud/QX7dz9zhSE2m.128.mp3 -> /home/cdamian/Music/gpodder-downloads/motik on Soundcloud/Motik 121212121212 mp3.mp3
1623063623.171142 [gpodder.dbsqlite] DEBUG: Commit.
1623063623.210063 [tagging] INFO: tagging.on_episode_downloaded(motik on Soundcloud/Motik 121212121212 mp3)
1623063623.210263 [gpodder.download] INFO: No more tasks for Thread-44 to carry out.
1623063624.380371 [gpodder.gtkui.model] DEBUG: cached thumb wrong size: (32, 32) != 34
1623063624.380828 [gpodder.gtkui.main] INFO: All tasks have finished.
1623063624.577726 [gpodder.gtkui.model] DEBUG: cached thumb wrong size: (32, 32) != 34
1623063624.578174 [gpodder.gtkui.main] INFO: All tasks have finished.




1623063656.567940 [gpodder.util] DEBUG: run_in_background: <function gPodder.update_feed_cache.<locals>.update_feed_cache_proc at 0x7ff31447c9d0> (False)
1623063656.571794 [gpodder.registry] INFO: feed_handler resolved by fetch_channel from gpodder.plugins.soundcloud: None -> <gpodder.feedcore.Result object at 0x7ff3144daa90>
1623063656.572158 [gpodder.plugins.soundcloud] DEBUG: loading https://api.soundcloud.com/users/901720/tracks.json?consumer_key=zrweghtEtnZLpXf3mlm8mQ&limit=200
1623063656.573505 [urllib3.connectionpool] DEBUG: Starting new HTTPS connection (1): api.soundcloud.com:443
1623063656.615065 [gpodder.gtkui.main] INFO: Updating motik on Soundcloud (1/1)
1623063656.826620 [urllib3.connectionpool] DEBUG: https://api.soundcloud.com:443 "GET /users/901720/tracks.json?consumer_key=zrweghtEtnZLpXf3mlm8mQ&limit=200 HTTP/1.1" 200 1604
1623063656.827936 [gpodder.plugins.soundcloud] INFO: 2/2 downloadable tracks for user motik tracks feed
1623063656.843570 [gpodder.dbsqlite] DEBUG: Purge requested for podcast 19
1623063656.843856 [gpodder.dbsqlite] DEBUG: Commit.
1623063656.855410 [gpodder.gtkui.services] DEBUG: cover download request for http://soundcloud.com/motik
1623063656.855485 [gpodder.util] DEBUG: run_in_background: <function CoverDownloader.request_cover.<locals>.<lambda> at 0x7ff3177a7040> (False)
1623063656.855958 [gpodder.coverart] INFO: Downloading cover art: https://i1.sndcdn.com/avatars-000001146600-xeho16-large.jpg
1623063656.857673 [urllib3.connectionpool] DEBUG: Starting new HTTPS connection (1): i1.sndcdn.com:443
1623063656.867819 [gpodder.gtkui.main] DEBUG: Updated channel is active, updating UI
1623063656.868362 [gpodder.my] DEBUG: Processing received episode actions...
1623063656.873104 [gpodder.my] DEBUG: Received episode actions processed.
1623063656.873219 [gpodder.dbsqlite] DEBUG: Commit.
1623063656.986377 [urllib3.connectionpool] DEBUG: https://i1.sndcdn.com:443 "GET /avatars-000001146600-xeho16-large.jpg HTTP/1.1" 404 21
1623063656.986873 [gpodder.coverart] WARNING: Cover art download failed: https://i1.sndcdn.com/avatars-000001146600-xeho16-large.jpg returned status code 404
1623063656.988794 [gpodder.gtkui.model] INFO: Clearing cover from cache: http://soundcloud.com/motik

```

**Comment by auouymous** [comment]
Updated today and it removed the two deleted motik episodes and added them again. One was marked new, the other old. the following error was thrown, for each episode.

```
1623124061.230031 [gpodder.dbsqlite] ERROR: Cannot save <gpodder.gtkui.model.GEpisode object at 0xab5ac5d8>: UNIQUE constraint failed: episode.podcast_id, episode.guid
Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/gpodder/dbsqlite.py", line 214, in _save_object
    cur.execute(sql, values)
sqlite3.IntegrityError: UNIQUE constraint failed: episode.podcast_id, episode.guid
```

Restarting gpodder allows an update to reproduce the issue immediately. I restarted my computer between yesterday and today which is probably why I got the error today.

One problem is that `existing_guids` has string keys and `track['guid']` returns an int which never matches the keys (maybe it worked in an old python). This causes all episodes to be added on each update and is the reason for the non-unique guid collisions above.

```patch
-            if track['guid'] not in existing_guids:
+            if str(track['guid']) not in existing_guids:
```

That fixes the issue but something else is then removing the existing episodes from the database...

**Comment by auouymous** [comment]
#1065 appears to work.

@christofdamian Do you know how to apply a patch or hand edit gpodder's python files? If so, could you install https://github.com/gpodder/gpodder/pull/1065.patch and let me know if it fixes the issue?

**Comment by christofdamian** [comment]
@auouymous I tried your branch and it does seem to have fixed it. On first run it downloaded every available podcast, but the second run just got the new ones.

**Comment by christofdamian** [comment]
Definitely working fine now. Thanks for the fix. 
Fixed by: #1065

**Comment by auouymous** [comment]
Thanks for testing.
